### ISSUE TITLE
[Bug] Loss of TrueNAS connection results in any further PVC creations to fail due to already provisioned

### ISSUE THREAD

**Issue by 0xEmma** [democratic-csi/democratic-csi#351]
In the event of a truenas reboot/changing network settings, the controller will be unable to make any further PVC's, and result in the following error:
```failed to provision volume with StorageClass "freenas-nfs-csi": rpc error: code = AlreadyExists desc = volume has already been created with a different size, existing size: -, required_bytes: 5368709120, limit_bytes: 0```

Simply restarting the controller pod results in a sucess
Using the FreeNAS-API-NFS setup.

**Comment by travisghansen** [comment]
That seems pretty strange. The calls are idempotent so I wonder what else could have put the controller in that state :(

**Comment by 0xEmma** [comment]
Seem to get the same error + aswell as the truenas gui yelling at me for invalid quotas when setting up a PVC when the system is under high load aswell, only makes the PVC with no quotas or NFS share

**Comment by oscgu** [comment]
Also hit the `AlreadyExists` issue, but restarting the controller pod did fix it. I deleted a pv/c and then tried to re-create it, which then lead to the creation of the new one to fail. We're using TrueNAS-Scale api iscsi setup.

**Comment by inzanez** [comment]
Well, same here, thanks a lot for the restart tip!! I would have spent hours otherwise,...:-)

**Comment by travisghansen** [comment]
If someone can send me the logs from the controller when that happens when the api logs are enabled that would be great. The controller is stateless so restarting it to fix the issue seems really odd.

**Comment by d-uzlov** [comment]
I have logs with `log-level=debug`, I hope this is enough. They are from a somewhat older version `docker.io/democraticcsi/democratic-csi@sha256:945cf404d40f719edda62c1889d142bfe8b970c7bfd3e99054869e1dc8efe62a`. I can't quite reproduce it on purpose, so I can't immediately test it with the latest version.
I will post logs from the latest version if I encounter this bug again.

```log
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"info","message":"new request - driver: FreeNASApiDriver method: CreateVolume call: {\"metadata\":{\"x-forwarded-host\":[\"localhost\"],\"user-agent\":[\"grpc-go/1.54.0\"]},\"request\":{\"volume_capabilities\":[{\"access_mode\":{\"mode\":\"MULTI_NODE_MULTI_WRITER\"},\"mount\":{\"mount_flags\":[\"soft\",\"noatime\",\"rsize=262144\",\"wsize=262144\"],\"fs_type\":\"nfs\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}],\"parameters\":{\"detachedVolumesFromSnapshots\":\"false\",\"fsType\":\"nfs\",\"csi.storage.k8s.io/pvc/name\":\"test-nfs-bulk\",\"csi.storage.k8s.io/pvc/namespace\":\"default\",\"csi.storage.k8s.io/pv/name\":\"pvc-c1521a07-8d20-4a49-8124-084f34b3c26e\"},\"secrets\":\"redacted\",\"name\":\"pvc-c1521a07-8d20-4a49-8124-084f34b3c26e\",\"capacity_range\":{\"required_bytes\":\"104857600\",\"limit_bytes\":\"0\"},\"volume_content_source\":null,\"accessibility_requirements\":null},\"cancelled\":false}","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.011Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"debug","message":"operation lock keys: [\"create_volume_name_pvc-c1521a07-8d20-4a49-8124-084f34b3c26e\"]","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.011Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"verbose","message":"validating capabilities: [{\"access_mode\":{\"mode\":\"MULTI_NODE_MULTI_WRITER\"},\"mount\":{\"mount_flags\":[\"soft\",\"noatime\",\"rsize=262144\",\"wsize=262144\"],\"fs_type\":\"nfs\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}]","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.016Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"debug","message":"FREENAS HTTP REQUEST: {\"headers\":{\"Accept\":\"application/json\",\"User-Agent\":\"democratic-csi-driver\",\"Content-Type\":\"application/json\",\"Authorization\":\"redacted\"},\"responseType\":\"json\",\"timeout\":60000,\"method\":\"GET\",\"url\":\"http://hdd.tn.lan/api/v1.0/pool/dataset/id/main%2Fk8s%2Fnfsd%2Fdefault--test-nfs-bulk/\"}","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.035Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"debug","message":"FREENAS HTTP ERROR: null","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.035Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"debug","message":"FREENAS HTTP STATUS: 200","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.035Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"debug","message":"FREENAS HTTP HEADERS: {\"server\":\"nginx\",\"date\":\"Thu, 01 Feb 2024 19:01:48 GMT\",\"content-type\":\"text/html\",\"last-modified\":\"Tue, 25 Jul 2023 00:04:41 GMT\",\"transfer-encoding\":\"chunked\",\"connection\":\"keep-alive\",\"cache-control\":\"must-revalidate\",\"strict-transport-security\":\"max-age=0; includeSubDomains; preload\",\"x-content-type-options\":\"nosniff\",\"x-xss-protection\":\"1; mode=block\",\"permissions-policy\":\"geolocation=(),midi=(),sync-xhr=(),microphone=(),camera=(),magnetometer=(),gyroscope=(),fullscreen=(self),payment=()\",\"referrer-policy\":\"strict-origin\",\"x-frame-options\":\"SAMEORIGIN\"}","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.035Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"debug","message":"FREENAS HTTP BODY: \"<!DOCTYPE html><html lang=\\\"en\\\"><head>\\n  <meta charset=\\\"utf-8\\\">\\n  <meta http-equiv=\\\"Pragma\\\" content=\\\"no-cache\\\">\\n  <meta http-equiv=\\\"Expires\\\" content=\\\"-1\\\">\\n  <meta http-equiv=\\\"CACHE-CONTROL\\\" content=\\\"NO-CACHE\\\">\\n  <script nomodule src=\\\"assets/scripts/ie-support/ie-polyfills.min.js\\\"></script>\\n  <title id=\\\"main-page-title\\\"></title>\\n  <base href=\\\"/ui/\\\">\\n  <meta name=\\\"viewport\\\" content=\\\"width=device-width, initial-scale=1\\\">\\n  <link rel=\\\"apple-touch-icon\\\" sizes=\\\"120x120\\\" href=\\\"assets/favicons/apple-touch-icon.png\\\">\\n  <link rel=\\\"icon\\\" type=\\\"image/png\\\" sizes=\\\"32x32\\\" href=\\\"assets/favicons/favicon-32x32.png\\\">\\n  <link rel=\\\"icon\\\" type=\\\"image/png\\\" sizes=\\\"16x16\\\" href=\\\"assets/favicons/favicon-16x16.png\\\">\\n  <link rel=\\\"manifest\\\" href=\\\"assets/favicons/site.webmanifest\\\">\\n  <link rel=\\\"mask-icon\\\" href=\\\"assets/favicons/safari-pinned-tab.svg\\\" color=\\\"#000000\\\">\\n  <link rel=\\\"shortcut icon\\\" href=\\\"assets/favicons/favicon.ico\\\">\\n  <meta name=\\\"msapplication-TileColor\\\" content=\\\"#2d89ef\\\">\\n  <meta name=\\\"msapplication-config\\\" content=\\\"assets/favicons/browserconfig.xml\\\">\\n  <meta name=\\\"theme-color\\\" content=\\\"#ffffff\\\">\\n  <style>\\n    :root {\\n      --bg1: #1e1e1e;\\n    }\\n\\n    html {\\n      background: var(--bg1);\\n    }\\n\\n    @keyframes loading {\\n      50% {\\n        opacity: 0;\\n      }\\n    }\\n  </style>\\n<style>@charset \\\"UTF-8\\\";@-webkit-keyframes mdi-spin{0%{-webkit-transform:rotate(0deg);transform:rotate(0)}to{-webkit-transform:rotate(359deg);transform:rotate(359deg)}}:root{--fn-dropshadow:0 1px 2px rgba(0, 0, 0, 0);--contrast-normal:var(--bg2);--contrast-darker:hsl(0, 0%, 10.7%);--contrast-darkest:hsl(0, 0%, 5.7%);--contrast-lighter:hsl(0, 0%, 20.7%);--contrast-lightest:hsl(0, 0%, 25.7%);--light-theme-lines:var(--contrast-darkest);--dark-theme-lines:var(--contrast-lighter);--lines:var(--dark-theme-lines);--sidenav-width:240px;--btn-default-bg:var(--alt-bg2);--hover-bg:var(--lines) !important;--primary-lighter:var(--primary);--font-family-header:\\\"Titillium Web\\\", sans-serif;--font-family-body:\\\"IBM Plex Sans\\\", sans-serif}@font-face{font-display:swap;font-family:Roboto;font-style:italic;font-weight:300;src:local(\\\"Roboto Light Italic\\\"),local(\\\"Roboto-LightItalic\\\"),url(Roboto-LightItalic.1808ee290b327719.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:italic;font-weight:400;src:local(\\\"Roboto Italic\\\"),local(\\\"Roboto-Italic\\\"),url(Roboto-Italic.110d959f7d4d7b91.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:italic;font-weight:500;src:local(\\\"Roboto Medium Italic\\\"),local(\\\"Roboto-MediumItalic\\\"),url(Roboto-MediumItalic.732cb384d301906d.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:italic;font-weight:700;src:local(\\\"Roboto Bold Italic\\\"),local(\\\"Roboto-BoldItalic\\\"),url(Roboto-BoldItalic.51813ec2952a0002.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:italic;font-weight:900;src:local(\\\"Roboto Black Italic\\\"),local(\\\"Roboto-BlackItalic\\\"),url(Roboto-BlackItalic.a09ba90185e40bcd.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:normal;font-weight:100;src:local(\\\"Roboto Thin\\\"),local(\\\"Roboto-Thin\\\"),url(Roboto-Thin.eb6aa271d5d6563c.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:normal;font-weight:300;src:local(\\\"Roboto Light\\\"),local(\\\"Roboto-Light\\\"),url(Roboto-Light.fdd4b0ffaa0a5e70.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:normal;font-weight:400;src:local(\\\"Roboto\\\"),local(\\\"Roboto-Regular\\\"),url(Roboto-Regular.fb80880011eeba42.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:normal;font-weight:500;src:local(\\\"Roboto Medium\\\"),local(\\\"Roboto-Medium\\\"),url(Roboto-Medium.d8dc431bbea2448e.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:normal;font-weight:700;src:local(\\\"Roboto Bold\\\"),local(\\\"Roboto-Bold\\\"),url(Roboto-Bold.5931183d69264abb.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:normal;font-weight:900;src:local(\\\"Roboto Black\\\"),local(\\\"Roboto-Black\\\"),url(Roboto-Black.1e6305dc809eced6.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Roboto;font-style:italic;font-weight:100;src:local(\\\"Roboto Thin Italic\\\"),local(\\\"Roboto-ThinItalic\\\"),url(Roboto-ThinItalic.e6a9459905d3e55e.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:IBM Plex Sans;src:url(IBMPlexSans-Regular.ceb4bac10177b34b.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:normal;font-weight:900;src:local(\\\"Titillium Web Black\\\"),local(\\\"TitilliumWeb-Black\\\"),url(TitilliumWeb-Black.05617876bd4e4548.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:normal;font-weight:700;src:local(\\\"Titillium Web Bold\\\"),local(\\\"TitilliumWeb-Bold\\\"),url(TitilliumWeb-Bold.63d205ade976fd06.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:italic;font-weight:700;src:local(\\\"Titillium Web Bold Italic\\\"),local(\\\"TitilliumWeb-BoldItalic\\\"),url(TitilliumWeb-BoldItalic.01c852359543683a.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:italic;font-weight:400;src:local(\\\"Titillium Web Italic\\\"),local(\\\"TitilliumWeb-Italic\\\"),url(TitilliumWeb-Italic.8322b4b31b8fd0fa.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:normal;font-weight:300;src:local(\\\"Titillium Web Light\\\"),local(\\\"TitilliumWeb-Light\\\"),url(TitilliumWeb-Light.27420d5c2534048b.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:italic;font-weight:300;src:local(\\\"Titillium Web Light Italic\\\"),local(\\\"TitilliumWeb-LightItalic\\\"),url(TitilliumWeb-LightItalic.a0b816123392eb3b.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:normal;font-weight:400;src:local(\\\"Titillium Web\\\"),local(\\\"TitilliumWeb-Regular\\\"),url(TitilliumWeb-Regular.54796315af076ab7.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:normal;font-weight:500;src:local(\\\"Titillium Web SemiBold\\\"),local(\\\"TitilliumWeb-SemiBold\\\"),url(TitilliumWeb-SemiBold.61c7648cb3623785.ttf) format(\\\"truetype\\\")}@font-face{font-display:swap;font-family:Titillium Web;font-style:italic;font-weight:500;src:local(\\\"Titillium Web SemiBold Italic\\\"),local(\\\"TitilliumWeb-SemiBoldItalic\\\"),url(TitilliumWeb-SemiBoldItalic.ad90425aa90c02de.ttf) format(\\\"truetype\\\")}html{font-size:87.5%}html,body{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;height:100%;min-height:100%;position:relative;-webkit-tap-highlight-color:transparent;-webkit-text-size-adjust:100%;-moz-text-size-adjust:100%;text-size-adjust:100%;-webkit-touch-callout:none;width:100%}body{background:var(--bg1);color:#000000de;font-family:IBM Plex Sans,Roboto,Helvetica Neue,sans-serif;font-size:.875rem;font-weight:400;line-height:1.65;margin:0;padding:0}div{box-sizing:border-box}@-webkit-keyframes silde_to_top{0%{bottom:-60px}to{bottom:0;z-index:1000000}}</style><link rel=\\\"stylesheet\\\" href=\\\"styles.737d2e71e0ad6957.css\\\" media=\\\"print\\\" onload=\\\"this.media='all'\\\"><noscript><link rel=\\\"stylesheet\\\" href=\\\"styles.737d2e71e0ad6957.css\\\"></noscript></head>\\n<body class=\\\"ix-dark\\\">\\n  <ix-root>\\n    <div style=\\\"display: flex; align-items: center; justify-content: center; height: 100vh;\\\">\\n      <img src=\\\"assets/customicons/truenas_scale_logo_full.svg\\\" style=\\\"animation: loading 2s linear infinite; filter: invert(1); opacity: 0.1\\\" width=\\\"300\\\" height=\\\"65\\\" alt=\\\"TrueNAS SCALE\\\">\\n    </div>\\n  </ix-root>\\n<script src=\\\"runtime.436badb8355fec6f.js\\\" type=\\\"module\\\"></script><script src=\\\"polyfills.5c83906d166a3723.js\\\" type=\\\"module\\\"></script><script src=\\\"scripts.f70c08e53d3e7caa.js\\\" defer></script><script src=\\\"main.6770cb0f6c323351.js\\\" type=\\\"module\\\"></script>\\n\\n</body></html>\"","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.035Z"}
{"host":"dnfsb-controller-7dbc8c55d9-867lg","level":"error","message":"handler error - driver: FreeNASApiDriver method: CreateVolume error: {\"name\":\"GrpcError\",\"code\":6,\"message\":\"volume has already been created with a different size, existing size: -, required_bytes: 104857600, limit_bytes: 0\"}","service":"democratic-csi","timestamp":"2024-02-01T19:01:48.036Z"}
```

**Comment by travisghansen** [comment]
I may have an idea what is happening here. Thanks for sharing the logs! The controller at regular intervals checks the api endpoint to determine if it should use the v1 or v2 endpoints. It seems the v1 (very old and no longer relevant) endpoint is getting hit after going into a bad state. I’ll dig through the code and see if I can find a sane way to resolve this problem.

**Comment by travisghansen** [comment]
Should be fixed in this commit: https://github.com/democratic-csi/democratic-csi/commit/536f954e0a5e94a706f4332e530f6fab332a19e1

**Comment by travisghansen** [comment]
Fix should be released in `v1.9.0`. Reopen if you see the behavior continue after upgrading.
